### PR TITLE
Uk/1563 backend for persistent lang switching

### PR DESCRIPTION
**Pull request by jeronimo**
This is extension to PR https://github.com/openfoodfoundation/openfoodnetwork/pull/1564 and back-end to #1563, meaning that passing ?locale=en-GB would switch and persist by all requirements described in Slack #multilingual channel.

P.S. Somehow extracting login from I18nHelper#set_locale to user's model seemed too much, so left as it is to keep the logic in one file.

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -123,8 +123,8 @@
   create_table "distributors_shipping_methods", :force => true do |t|
     t.integer  "distributor_id"
     t.integer  "shipping_method_id"
-    t.datetime "created_at",         :null => false
-    t.datetime "updated_at",         :null => false
+    t.datetime "created_at"
+    t.datetime "updated_at"
```
Comment (enricostano) [inline db/schema.rb]: Is this related to this PR?

Comment (jeronimo) [inline db/schema.rb]: That's what had been generated by default by `rake db:migrate`

Comment (enricostano) [inline db/schema.rb]: We shouldn't run `rake db:migrate` every time in development because different versions of Rails could generate different schema. I suggest to just run `rake db:schema:load` to make sure your branch doesn't touch the schema if not needed.

If your branch adds new migrations just run `rake db:migrate:up VERSION={MIGRATION_ID}` and commit the changes that your migration introduce.

Does this make sense? The risk here is to mix commits and change the DB schema when it's not related with your PR.

Comment (jeronimo) [inline db/schema.rb]: Totaly! Thanks!

Comment (jeronimo) [inline db/schema.rb]: Had manually to unstage those changes. As `rake db:schema:load` throws PostgreSQL error. and `rake db:migrate:up VERSION={MIGRATION_ID}` still removes those `:null => false`. So those suggestions do not work.

Comment (enricostano) [inline db/schema.rb]: We do have PostgreSQL authentication issues in development :(

In any case, your branch doesn't have any migration. Why are you running `rake db:migrate` at all?

Comment (jeronimo) [inline db/schema.rb]: https://github.com/openfoodfoundation/openfoodnetwork/pull/1572/files#diff-cb64f2e8599ce0d293d36ffd24002014

Comment (enricostano) [inline db/schema.rb]: Sorry, I didn't remember this PR quite well. You do have a migration in this PR. `rake db:migrate:up VERSION=20170512115519` should work. Try to `down` it and then `up` again the specific migration.

Comment (jeronimo) [inline db/schema.rb]: Believe me I have tried :)

File: app/helpers/i18n_helper.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module I18nHelper
+
+  private
+  def set_locale
+    # update current_spree_user locale if logged in and set cookie locale if params
+    if params[:locale] && Rails.application.config.i18n.available_locales.include?(params[:locale])
+      current_spree_user.update_attributes!(locale: params[:locale]) if current_spree_user
```
Comment (oeoeaio) [inline app/helpers/i18n_helper.rb]: I'm sure `current_spree_user` and `spree_current_user` are the same thing (correct me if that is not right). We have been using `spree_current_user` everywhere else in the code, so would be good to stick with that convention.

File: README.md [hunk 3]
```diff
@@ -107,6 +107,9 @@ If you want karma to automatically rerun the tests on file modification, use:
 
     ./script/karma start
 
+### Multilingual
+Do not forget to run `rake tmp:cache:clear` after locales are updated to reload I18n js translations.
```
Comment (oeoeaio) [inline README.md]: Does this need to be included in any deployment scripts, or is restarting the server sufficient?

Comment (jeronimo) [inline README.md]: @oeoeaio, Good question! I don't know. In dev environment it is not enough to restart the server. So I assume it won't be enough in production as well... Is there a way I could test it?

**Comment by daniellemoorhead** [comment]
@mkllnk could you have a look at this as suggested by @oeoeaio please :)

**Comment by Matt-Yorkley** [comment]
The language switcher is looking good! I'll be working on this issue this week and looking at implementing the frontend.

File: app/helpers/i18n_helper.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+module I18nHelper
+
+  private
```
Comment (sauloperez) [inline app/helpers/i18n_helper.rb]: `private` has no effect in modules and besides I don't really understand what you are trying to achive with it.

Comment (mkllnk) [inline app/helpers/i18n_helper.rb]: This got merged in an earlier PR already, but I would be happy to remove it in this PR.

**Comment by mkllnk** [comment]
@sauloperez Do you think we should add a helper spec here?

I resolved the conflicts by rebasing and omitting the commits that had been merged to master already.

File: app/helpers/i18n_helper.rb [hunk 5]
```diff
@@ -1,13 +1,12 @@
 module I18nHelper
-  private
```
Comment (sauloperez) [inline app/helpers/i18n_helper.rb]: thanks!

**Comment by sauloperez** [comment]
> @sauloperez Do you think we should add a helper spec here?

Yes, I would.

**Comment by mkllnk** [comment]
@sauloperez Can you review my added specs?

File: spec/helpers/i18n_helper_spec.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
+require 'spec_helper'
+
+describe I18nHelper do
+  let(:user) { create(:user) }
+
+  context "as guest" do
+    before do
+      allow(helper).to receive(:spree_current_user) { nil }
```
Comment (sauloperez) [inline spec/helpers/i18n_helper_spec.rb]: where is `helper` defined?

Comment (enricostano) [reply spec/helpers/i18n_helper_spec.rb]: Rspec/Rails magic

Comment (sauloperez) [reply spec/helpers/i18n_helper_spec.rb]: I couldn't find it Rails guides. That's why I ask. Perhaps is something from RSpec.

Comment (enricostano) [reply spec/helpers/i18n_helper_spec.rb]: https://github.com/rspec/rspec-rails#helper-specs

**Comment by mkllnk** [comment]
The specs still need a bit of work. Travis fails reliably. I think it depends on the order of specs. I can reproduce the issue with:

```sh
export CI_NODE_TOTAL=5
export CI_NODE_INDEX=0
bundle exec rake 'knapsack:rspec[--tag ~performance]'
```
```
  1) Enterprise associations ownership validates ownership limit                                                                            
     Failure/Error: expect{
       expected ActiveRecord::RecordInvalid with "Validation failed: laurine@nolangottlieb.uk is not permitted to own any more enterprises (limit is 1).", got #<ActiveRecord::RecordInvalid: La validación falló: Owner translation missing: es.enterprise_owner_error> with backtrace:
         # ./spec/models/enterprise_spec.rb:174:in `block (5 levels) in <top (required)>'
         # ./spec/models/enterprise_spec.rb:172:in `block (4 levels) in <top (required)>'
     # ./spec/models/enterprise_spec.rb:172:in `block (4 levels) in <top (required)>'

  2) Enterprise validations requires an owner                                                                                               
     Failure/Error: expect{
       expected ActiveRecord::RecordInvalid with "Validation failed: Owner can't be blank", got #<ActiveRecord::RecordInvalid: La validación falló: Owner no puede estar en blanco> with backtrace:
         # ./spec/models/enterprise_spec.rb:188:in `block (4 levels) in <top (required)>'
         # ./spec/models/enterprise_spec.rb:187:in `block (3 levels) in <top (required)>'
     # ./spec/models/enterprise_spec.rb:187:in `block (3 levels) in <top (required)>'

  3) Enterprise validations name uniqueness prevents duplicate names for new records                                                        
     Failure/Error: "has already been taken. If this is your enterprise and you would like to claim ownership, please contact the current manager of this profile at owner@example.com."
       expected: "has already been taken. If this is your enterprise and you would like to claim ownership, please contact the current manager of this profile at owner@example.com."
            got: "translation missing: es.enterprise_name_error" (using ==)
     # ./spec/models/enterprise_spec.rb:200:in `block (4 levels) in <top (required)>'

  4) Enterprise validations name uniqueness prevents duplicate names for existing records                                                   
     Failure/Error: "has already been taken. If this is your enterprise and you would like to claim ownership, please contact the current manager of this profile at owner@example.com."
       expected: "has already been taken. If this is your enterprise and you would like to claim ownership, please contact the current manager of this profile at owner@example.com."
            got: "translation missing: es.enterprise_name_error" (using ==)
     # ./spec/models/enterprise_spec.rb:208:in `block (4 levels) in <top (required)>'

  5) destruction as the last variant of a product cannot be deleted                                                                         
     Failure/Error: expect(first_variant.errors[:product]).to include "must have at least one variant"
       expected ["translation missing: es.spree_variant_product_error"] to include "must have at least one variant"
     # ./spec/models/spree/variant_spec.rb:538:in `block (3 levels) in <module:Spree>'
```

It looks like the language has been set to Spanish in a spec and that persists.

**Comment by sstead** [comment]
@mkllnk Testing looks good
